## 创建环境，启动与说明（必看） ：

由于本次作业时作业有关data的作业讲解，为了简单易学，这里使用的都不是全量数据集，一是因为数据集很大，下载非常耗时，并且考虑到内存和模型训练的时间，这里使用部分数据集，只下载一部分进行使用，如果读者想要下载完整数据集可以下载，流程都是共通的。

作者使用的是 assignmnet2 system 中wsl的ubuntu远程连接vs code，操作参考assignmnet2。

在终端中运行：

```bash
uv sync
. venv/bin/activate
```

# 作业一

**问题（look_at_cc）：4 分**

(a) 下载上面的 WARC 文件，或者找到我们在集群上提供的副本。让我们看看这个文件中的第一页。这是一个 gzipped 文件，你可以使用以下命令浏览其内容：

```
$ zcat /data/CC/example.warc.gz | less
```

`less` 让你可以使用键盘箭头、Page Up、Page Down 来浏览文件。要退出，按“q”。
查看非常第一个网页。它的 URL 是什么？它仍然可以访问吗？你能通过查看原始 HTML 来告诉我们这个页面似乎是什么吗？
交付物：2-3 句回应。

(b) 现在让我们看看对应的 WET 文件：

```
$ zcat /data/CC/example.warc.wet.gz | less
```

注意 WET 文件包含 HTTP 头（例如，Content-Length），这些不是提取的文本内容的一部分。如果你查看第一个例子，你会看到它包含从你刚刚看到的原始 HTML 中提取的文本。

注意提取的文本中有多少是 HTML 结构的遗迹，而不是页面的主要内容。你认为哪些部分应该被提取器过滤掉？将此文本作为训练数据时，可能会出错的地方是什么：在训练看起来像这样的文本的模型时，可能会出错的地方是什么？相反，模型可能从这个页面中提取哪些有用的信息？
交付物：3-4 句回应。

(c) 什么使一个好的训练示例具有高度的上下文性。描述一个应用领域，在这个领域中，这个示例可能对训练数据有用，以及一个它可能没有用的地方。
交付物：1-2 句回应。

(d) 让我们看更多的例子，以便更好地了解 Common Crawl 中的内容。浏览另外 25 个 WET 记录。对于每条记录，非常简要地评论文档的语言（如果可以识别），域名，页面类型等。你需要看到多少个例子才能确定你认为的“高质量”网页？
交付物：对 25 个文档进行简短注释，包括文档的语言、域名、页面类型和其他任何关于文档的杂项注释。直到你看到一个高质量的示例。示例的数量取决于你看到高质量示例的时间。


运行以下命令来下载数据集
```bash
wget https://data.commoncrawl.org/crawl-data/CC-MAIN-2025-18/segments/1744889135610.12/warc/CC-MAIN-20250417135010-20250417165010-00065.warc.gz

wget https://data.commoncrawl.org/crawl-data/CC-MAIN-2025-18/segments/1744889135610.12/wet/CC-MAIN-20250417135010-20250417165010-00065.warc.wet.gz
```

In [1]:
!wget --no-proxy https://data.commoncrawl.org/crawl-data/CC-MAIN-2025-18/segments/1744889135610.12/warc/CC-MAIN-20250417135010-20250417165010-00065.warc.gz


--2026-01-26 13:23:52--  https://data.commoncrawl.org/crawl-data/CC-MAIN-2025-18/segments/1744889135610.12/warc/CC-MAIN-20250417135010-20250417165010-00065.warc.gz
Resolving data.commoncrawl.org (data.commoncrawl.org)... 18.238.217.12, 18.238.217.90, 18.238.217.75, ...
Connecting to data.commoncrawl.org (data.commoncrawl.org)|18.238.217.12|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1121135968 (1.0G) [application/octet-stream]
Saving to: ‘CC-MAIN-20250417135010-20250417165010-00065.warc.gz’

CC-MAIN-20250417135 100%[===================>]   1.04G  6.10MB/s    in 2m 58s  

2026-01-26 13:27:04 (6.00 MB/s) - ‘CC-MAIN-20250417135010-20250417165010-00065.warc.gz’ saved [1121135968/1121135968]



In [2]:

!wget --no-proxy  https://data.commoncrawl.org/crawl-data/CC-MAIN-2025-18/segments/1744889135610.12/wet/CC-MAIN-20250417135010-20250417165010-00065.warc.wet.gz

--2026-01-26 13:27:04--  https://data.commoncrawl.org/crawl-data/CC-MAIN-2025-18/segments/1744889135610.12/wet/CC-MAIN-20250417135010-20250417165010-00065.warc.wet.gz
Resolving data.commoncrawl.org (data.commoncrawl.org)... 3.163.189.35, 3.163.189.100, 3.163.189.118, ...
Connecting to data.commoncrawl.org (data.commoncrawl.org)|3.163.189.35|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 80904812 (77M) [application/octet-stream]
Saving to: ‘CC-MAIN-20250417135010-20250417165010-00065.warc.wet.gz’

CC-MAIN-20250417135 100%[===================>]  77.16M  6.09MB/s    in 14s     

2026-01-26 13:27:21 (5.47 MB/s) - ‘CC-MAIN-20250417135010-20250417165010-00065.warc.wet.gz’ saved [80904812/80904812]



我们可以使用
```bash
zcat CC-MAIN-20250417135010-20250417165010-00065.warc.gz | less

zcat CC-MAIN-20250417135010-20250417165010-00065.warc.wet.gz | less
```
在命令行中进行预览，less表示可以使用键盘方向键、Page Up 和 Page Down 来浏览文件。要退出，请按“q“。


`.warc.gz` 和 `.warc.wet.gz` 都是 **WARC（Web ARChive）格式** 的压缩文件，广泛用于大规模网页抓取数据（如 Common Crawl）。

## 一、`.warc.gz` 与 `.warc.wet.gz` 的区别

### 1. `.warc.gz`

**原始网页归档文件**，包含完整抓取结果：HTTP 请求 / 响应头、HTML 原文、二进制资源（图片、PDF 等）、元数据（抓取时间、URL、状态码等）。

### 2. `.warc.wet.gz`

**WET = Web Extracted Text**

从 `.warc.gz` 中**抽取后的纯文本**，仅保留网页的可读文本内容，已去除 HTML 标签、脚本、样式等

适合NLP 训练（LM、Embedding、Topic），文本挖掘，语料构建。




In [3]:
# 从 warcio 库中导入 ArchiveIterator
# ArchiveIterator 用于按“记录（record）”顺序流式读取 WARC 文件
from warcio.archiveiterator import ArchiveIterator

# 导入 gzip，用于读取 .gz 压缩格式的 WARC 文件
import gzip

# 导入 html 模块，用于将 HTML 实体（如 &#x4EBA;）解码为正常字符
import html 

# WARC 文件路径（Common Crawl 的一个分片）
# 注意：变量名 flie_path 拼写不影响运行，但语义上建议改为 file_path
flie_path  = 'CC-MAIN-20250417135010-20250417165010-00065.warc.gz'

# 以二进制只读模式打开 gzip 压缩的 WARC 文件
# 必须使用 "rb"，因为 WARC 是二进制格式
with gzip.open(flie_path, "rb") as f:

    # 使用 ArchiveIterator 对 WARC 文件进行流式遍历
    # 每次循环返回一个 WARC record（请求、响应、元数据等）
    for step,record in enumerate(ArchiveIterator(f)):
       
        # 只处理类型为 "response" 的记录
        # response 表示真实的 HTTP 响应内容（网页、图片等）
        if record.rec_type == "response" and step > 5:

            # 从 WARC 头中获取该网页对应的原始 URL
            url = record.rec_headers.get_header("WARC-Target-URI")

            # 读取 HTTP 响应体内容（网页原始字节流）
            # decode 为 utf-8，忽略非法字符，防止程序崩溃
            html1 = record.content_stream().read().decode("utf-8", errors="ignore")

            # 简单判断该响应是否为 HTML 页面
            # 这是一个经验性过滤，避免处理图片、PDF 等非 HTML 内容
            if "<html" in html1.lower():

                # 输出网页 URL
                print(url)

                # 对 HTML 内容前 500 个字符进行：
                # 1. 截断（避免输出过多内容）
                # 2. HTML 实体解码（&#x4EBA; → 人）
                print(html.unescape(html1[:500]))

                # 找到第一个 HTML 页面后立即退出循环
            
                break


http://13.usnccm.org/
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML+RDFa 1.0//EN"
  "http://www.w3.org/MarkUp/DTD/xhtml-rdfa-1.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" version="XHTML+RDFa 1.0" dir="ltr"
  xmlns:content="http://purl.org/rss/1.0/modules/content/"
  xmlns:dc="http://purl.org/dc/terms/"
  xmlns:foaf="http://xmlns.com/foaf/0.1/"
  xmlns:og="http://ogp.me/ns#"
  xmlns:rdfs="http://www.w3.org/2000/01/rdf-schema#"
  xmlns:sioc="http://rdfs.org/sioc/ns#"
  xmlns:sioct="http://rdfs.org/sioc/ty


In [4]:
# 从 warcio 库中导入 ArchiveIterator
# ArchiveIterator 用于按顺序流式遍历 WARC / WET 文件中的每一条记录
from warcio.archiveiterator import ArchiveIterator

# 导入 gzip 模块，用于读取 .gz 压缩格式的文件
import gzip

# 指定 Common Crawl 的 WET 文件路径
# .warc.wet.gz 文件中存放的是已抽取好的“纯文本内容”
flie_path  = 'CC-MAIN-20250417135010-20250417165010-00065.warc.wet.gz'

# 以二进制只读模式打开 gzip 压缩的 WET 文件
# 必须使用 "rb"，因为 WARC/WET 属于二进制流格式
with gzip.open(flie_path, "rb") as f:

    # 使用 ArchiveIterator 对 WET 文件进行流式迭代
    # 每次循环返回一个 WARC record（一条网页对应一条 record）
    for record in ArchiveIterator(f):

        # 只处理类型为 "conversion" 的记录
        # conversion 表示从 HTML 中抽取出的“可读纯文本”
        if record.rec_type == "conversion":

            # 从 WARC 头信息中获取该文本对应的原始网页 URL
            url = record.rec_headers.get_header("WARC-Target-URI")

            # 读取当前 record 的正文内容（纯文本字节流）
            # 解码为 UTF-8 字符串，忽略非法字符，防止解码异常
            text = record.content_stream().read().decode("utf-8", errors="ignore")

            # 判断文本在去除首尾空白字符后是否仍然有内容
            # 用于过滤空页面或无有效文本的网页
            if len(text.strip()) > 0:

                # 打印网页对应的 URL
                print(url)

                # 打印正文内容的前 400 - 500 个字符
                # 用于快速查看文本样本，避免一次性输出过多内容
                print(text[400:500])

                # 找到第一条有效文本记录后立即退出循环
                # 适用于“快速验证 WET 文件内容是否正常”的场景
                break


http://0371rykj.com/ipfhsb/34.html
> ip防護(hù)系列 >
產(chǎn)品詳情/ products details
恒溫恒濕試驗(yàn)箱
產(chǎn)品用途
恒溫恒濕試驗(yàn)箱是航空、汽車(chē)、家電、科研等領(


# 作业二

**问题（提取文本）：3分**

（a）编写一个函数，用于从包含原始 HTML 的字节字符串中提取文本。使用 `resiliparse.extract.html2text.extract_plain_text` 执行提取操作。该函数需要一个字符串作为输入，因此您先需要将字节字符串解码为 Unicode 字符串。请注意，输入的字节字符串可能并未采用 UTF-8 编码，因此您的函数应具备检测编码的能力，以防 UTF-8 解码失败。此外，Resiliparse 还提供 `resiliparse.parse.encoding.detect_encoding()`，这可能会有用。

**交付物**：一个函数，接收包含 HTML 的字节字符串，并返回提取出的文本字符串。实现适配器 `[run_extract_text_from_html_bytes]`，并确保其通过 uv 运行
`pytest -k test_extract_text_from_html_bytes`

（b）对单个 WARC 文件运行你的文本提取功能。将其输出与提取的内容进行比较，对应 WET 文件中的文本。你注意到哪些差异和/或相似之处？哪种提取方式似乎更好？

**可交付成果**：用您自定义函数提取的文本与 WET 文件中的提取文本进行比较和对比，撰写 2–3 句话的回应。

---


In [5]:
from resiliparse.extract.html2text import extract_plain_text
from resiliparse.parse.encoding import detect_encoding


def run_extract_text_from_html_bytes(html_bytes: bytes) -> str:
    """
    从包含 HTML 的字节字符串中提取纯文本。

    参数:
        html_bytes (bytes): 原始 HTML 字节串

    返回:
        str: 提取出的纯文本
    """
    # 1. 尝试直接用 UTF-8 解码（最快路径）
    try:
        html_text = html_bytes.decode("utf-8")
    except UnicodeDecodeError:
        # 2. UTF-8 失败时，使用 resiliparse 的编码检测
        encoding = detect_encoding(html_bytes)
        # 使用检测到的编码进行解码，忽略无法解码的字符
        html_text = html_bytes.decode(encoding, errors="ignore")

    # 3. 使用 Resiliparse 从 HTML 字符串中提取纯文本
    text = extract_plain_text(html_text)

    return text


In [6]:
import gzip
from warcio.archiveiterator import ArchiveIterator
file_path = "CC-MAIN-20250417135010-20250417165010-00065.warc.gz"
with gzip.open(file_path, "rb") as f:
    for record in ArchiveIterator(f):
        if record.rec_type == "response":
            # 只处理 HTML 页面
            content_type = record.http_headers.get_header("Content-Type", "")
            if "text/html" not in content_type:
                continue

            # 1. 读取网页的原始 HTML（bytes）
            html_bytes = record.content_stream().read()

            # 2. 使用函数提取文本
            text = run_extract_text_from_html_bytes(html_bytes)

            # 3. 查看结果
            if text.strip():
                print(record.rec_headers.get_header("WARC-Target-URI"))
                print(text[:500])
                break


http://0371rykj.com/ipfhsb/34.html
久久久久女人精品毛片,99久久精品无码一区二区毛片,被老外的又粗又大日出了水,一边吃奶一边哭乱抻又乱扭

        • <th id="gckmo"></th>
        • <ul id="gckmo"><center id="gckmo"></center></ul>
      •  
  
 
 
 
         
   
         
    
         
    
        恒溫恒濕試驗(yàn)箱
        
	在線(xiàn)咨詢(xún)
    
         
    
         
      淋雨試驗(yàn)箱 
     
        上海林頻儀器股份有限公司Shanghai Linpin Instrument Stock Co Ltd
         
     
        服務(wù)熱線(xiàn)：4000 662 888 
         手機(jī)咨詢(xún)：13818467052
        
    
         
    
         
     



# 作业三

问题（语言识别）：6 分

(a) 编写一个函数，该函数将接收一个 Unicode 字符串，并识别其中出现的主要语言。你的函数应返回一个元组，包含该语言的标识符以及一个介于 0 和 1 之间的分数，表示对该预测的置信度。

**交付物：** 一个执行语言识别的函数，返回其最可能的语言预测及相应得分。实现适配器 `[run_identify_language]`，并确保它能通过 uv 中运行 `pytest -k test_identify_language` 的两项测试。请注意，这些测试假定英文使用特定的字符串标识符（“en”），中文使用 “zh”，因此你的测试适配器应根据需要执行任何必要的重新映射操作。

---

(b) 语言模型在推理阶段的行为在很大程度上取决于其训练所用的数据。因此，数据过滤管道中的问题可能会导致下游出现一系列难题。你认为语言识别环节出现问题可能引发哪些具体问题？在风险更高的场景中（例如部署面向用户的产品时），你将如何有效应对并缓解这些问题？

**可交付成果：** 2 至 5 句话的回复。

---

(c) 将您的语言识别系统应用于从 WARC 文件中提取的文本（通过您先前实现的文本提取功能）。在 20 个随机示例中手动识别语言，并将您的标注与分类器的预测结果进行对比。报告任何分类错误。有多少比例的文档是英语？根据您的观察，使用何种合适的分类器置信度阈值进行过滤会比较有效？

**交付物：** 2 至 5 句话的回复。





网络上包含着用数千种语言编写的页面。然而，在大多数计算预算下，训练一种能够有效利用如此多样化数据并实现大规模应用的多语言模型，仍是一项极具挑战性的任务。因此，许多源自Common Crawl的数据集所涵盖的语言种类相对有限。
一个适用于此目的的实用库是 fastText https://fasttext.cc，它提供了高效的文本分类器。该库既支持基于您自己的数据训练分类器的基础设施，也内置了一系列预训练模型，其中包括用于语言识别的模型。它提供了高效的文本分类器。该库既支持基于您自己的数据训练分类器的基础设施，也内置了一系列预训练模型，其中包括用于语言识别的模型。您可以从https://fasttext.cc/docs/en/language-identification.html 下载 fastText 语言识别模型 

通常，语言过滤器会利用分类器给出的得分来决定是否保留某页面。请使用fastText语言识别分类器实现一个语言识别过滤器，该过滤器应能输出一个非负分数，以表示其预测的置信度。


In [ ]:
# 下载模型 速度快，准确度也略高，但文件大小为 126MB

!wget --no-proxy https://dl.fbaipublicfiles.com/fasttext/supervised-models/lid.176.bin


--2026-01-26 13:27:21--  https://dl.fbaipublicfiles.com/fasttext/supervised-models/lid.176.bin
Resolving dl.fbaipublicfiles.com (dl.fbaipublicfiles.com)... 13.35.185.19, 13.35.185.64, 13.35.185.53, ...
Connecting to dl.fbaipublicfiles.com (dl.fbaipublicfiles.com)|13.35.185.19|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 131266198 (125M) [application/octet-stream]
Saving to: ‘lid.176.bin’

lid.176.bin         100%[===================>] 125.18M  9.93MB/s    in 14s     

2026-01-26 13:27:35 (8.88 MB/s) - ‘lid.176.bin’ saved [131266198/131266198]



In [8]:
# 下载模型 是该模型的压缩版本，文件大小为 917kB。

!wget --no-proxy https://dl.fbaipublicfiles.com/fasttext/supervised-models/lid.176.ftz

--2026-01-26 13:27:36--  https://dl.fbaipublicfiles.com/fasttext/supervised-models/lid.176.ftz
Resolving dl.fbaipublicfiles.com (dl.fbaipublicfiles.com)... 3.163.189.14, 3.163.189.51, 3.163.189.108, ...
Connecting to dl.fbaipublicfiles.com (dl.fbaipublicfiles.com)|3.163.189.14|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 938013 (916K) [binary/octet-stream]
Saving to: ‘lid.176.ftz’

lid.176.ftz         100%[===================>] 916.03K   777KB/s    in 1.2s    

2026-01-26 13:27:38 (777 KB/s) - ‘lid.176.ftz’ saved [938013/938013]



In [ ]:

import fasttext
from functools import lru_cache

# fastText 语言识别模型路径

@lru_cache(maxsize=1)
def _load_model():
    """
    加载 fastText 语言识别模型（单例模式）

    使用 lru_cache 的原因：
    - fastText 模型体积较大（~126MB）
    - 避免在每次函数调用时重复加载模型
    - 提高整体运行效率，尤其在批量文本处理时
    """
    return fasttext.load_model(MODEL_PATH)


def run_identify_language(text: str):
    """
    识别输入 Unicode 字符串中的主要语言

    参数：
        text (str): 任意 Unicode 文本字符串

    返回：
        (language_id, confidence_score)
        - language_id: 语言标识符（如 "en", "zh"）
        - confidence_score: 介于 [0, 1] 的置信度分数

    注意：
    - 该函数是测试要求的“适配器函数”
    - 必须保证英文返回 "en"，中文返回 "zh"
    """

    # 对空字符串或只包含空白符的情况进行兜底处理
    # fastText 在这种情况下预测结果不可靠
    if not text or not text.strip():
        return ("unknown", 0.0)

    # 加载（或从缓存中获取）语言识别模型
    model = _load_model()

    # fastText 的 predict 接口：
    # - 返回 labels 和 scores 两个列表
    # - labels 形如 "__label__en"
    # - scores 为预测置信度
    labels, scores = model.predict(
        text.replace("\n", " "),  # 去除换行，避免影响模型判断
        k=1                        # 只取概率最高的一个语言
    )

    # 取出预测结果
    raw_label = labels[0]                 # "__label__en"
    confidence = float(scores[0])         # 置信度（0~1）

    # 去掉 fastText 特有的 "__label__" 前缀
    language = raw_label.replace("__label__", "")

    # 根据作业测试要求，对语言标签进行统一映射
    # fastText 可能返回 zh, zh-cn, zh-tw 等
    if language.startswith("zh"):
        language = "zh"
    elif language == "en":
        language = "en"
    # 其他语言保持原样（如 "fr", "de", "ja" 等）

    return (language, confidence)


if __name__ == "__main__":
    
    examples = [
        "This is a simple English sentence.",
        "这是一个用于测试的中文句子。",
        "Bonjour, comment allez-vous ?",
        "こんにちは、元気ですか？",
        "",
        "    "
    ]

    for text in examples:
        MODEL_PATH = "lid.176.bin"
        lang, score = run_identify_language(text)
        print('126MB版本：')
        print(f"文本: {repr(text)}")
        print(f"预测语言: {lang}, 置信度: {score:.4f}")
        print("-" * 50)


126MB版本：
文本: 'This is a simple English sentence.'
预测语言: en, 置信度: 0.9385
--------------------------------------------------
126MB版本：
文本: '这是一个用于测试的中文句子。'
预测语言: zh, 置信度: 1.0001
--------------------------------------------------
126MB版本：
文本: 'Bonjour, comment allez-vous ?'
预测语言: fr, 置信度: 0.9883
--------------------------------------------------
126MB版本：
文本: 'こんにちは、元気ですか？'
预测语言: ja, 置信度: 1.0000
--------------------------------------------------
126MB版本：
文本: ''
预测语言: unknown, 置信度: 0.0000
--------------------------------------------------
126MB版本：
文本: '    '
预测语言: unknown, 置信度: 0.0000
--------------------------------------------------


# 作业四

## 问题（屏蔽 PII）：3 分

### 1. 屏蔽电子邮件地址

编写一个函数来屏蔽电子邮件。该函数将接收一个字符串作为输入，并用字符串
**`"|||EMAIL_ADDRESS|||"`**
替换所有出现的电子邮件地址。要检测电子邮件地址，你可以查阅能够可靠完成此任务的正则表达式。

**可交付成果：**
一个函数，用于将给定字符串中的所有电子邮件地址替换为字符串 `"|||EMAIL_ADDRESS|||"`，并返回一个元组，其中包含新字符串以及被屏蔽的实例数量。实现适配器 **`[run_mask_emails]`**，并确保其通过
`uv run pytest -k test_mask_emails` 中的所有测试。

---

### 2. 屏蔽电话号码

编写一个函数来屏蔽电话号码。该函数将接收一个字符串作为输入，并用字符串
**`"|||PHONE_NUMBER|||"`**
替换所有出现的电话号码。要可靠地实现这一点可能极具挑战性，因为电话号码的书写形式可能极其多样。

你应尽量捕获**美国最常用的电话号码格式**，并能有效应对轻微的语法偏差。

**可交付成果：**
一个函数，用于将给定字符串中的电话号码替换为字符串 `"|||PHONE_NUMBER|||"`，并返回包含新字符串及被屏蔽的电话号码数量的元组。实现适配器 **`[run_mask_phone_numbers]`**，并确保其通过
`uv run pytest -k test_mask_phones`。

---

### 3. 屏蔽 IP 地址

编写一个函数来屏蔽 IP 地址。对于这个问题，只需关注 **IPv4 地址**（由四个介于 0 到 255 之间的数字组成，各数字之间用点分隔）。

你的函数将接收一个字符串作为输入，并将所有出现的 IP 地址替换为字符串
**`"|||IP_ADDRESS|||"`**。

**可交付成果：**
一个函数，用于将给定字符串中的 IPv4 地址替换为字符串 `"|||IP_ADDRESS|||"`，并返回一个元组，其中包含新字符串以及被掩码替换的实例数量。实现适配器 **`[run_mask_ips]`**，并确保其通过
`uv run pytest -k test_mask_ips`。

---

### 4. 分析题（2–5 句）

你认为在训练集中如果**天真地**应用这些过滤器，可能会在语言模型的下游环节引发哪些问题？你又该如何缓解这些问题？

**交付物：** 2 至 5 句话的回复。

---

### 5. 实验分析（2–5 句）

在从 WARC 文件中提取的文本上运行你的 PII 脱敏函数（通过你先前实现的文本提取功能）。随机选取 20 个已进行替换的示例进行检查，并列举一些**假阳性（false positives）**和**假阴性（false negatives）**的具体示例。

**交付物：** 2 至 5 句话的回复。




In [10]:
import re
from typing import Tuple

def run_mask_emails(text: str) -> Tuple[str, int]:
    """
    屏蔽字符串中的电子邮件地址。

    参数:
        text (str): 输入文本

    返回:
        (masked_text, count)
    """

    # 常见电子邮件地址的正则表达式
    email_pattern = re.compile(
        r'\b[a-zA-Z0-9._%+-]+'
        r'@'
        r'[a-zA-Z0-9.-]+'
        r'\.[a-zA-Z]{2,}\b'
    )

    # findall 用于统计匹配数量
    matches = email_pattern.findall(text)

    # sub 用于统一替换
    masked_text = email_pattern.sub("|||EMAIL_ADDRESS|||", text)

    return masked_text, len(matches)


def run_mask_phone_numbers(text: str) -> Tuple[str, int]:


    phone_pattern = re.compile(
        r'''
        \b
        (?:\+?1[\s.-]?)?          # 可选国家码
        (?:\(?\d{3}\)?[\s.-]?)    # 区号
        \d{3}[\s.-]?\d{4}         # 主号码
        \b
        ''',
        re.VERBOSE
    )

    matches = phone_pattern.findall(text)
    masked_text = phone_pattern.sub("|||PHONE_NUMBER|||", text)

    return masked_text, len(matches)

def run_mask_ips(text: str) -> Tuple[str, int]:
    """
    屏蔽 IPv4 地址
    """

    ip_pattern = re.compile(
        r'\b'
        r'(?:25[0-5]|2[0-4]\d|1?\d{1,2})'
        r'(?:\.'
        r'(?:25[0-5]|2[0-4]\d|1?\d{1,2})){3}'
        r'\b'
    )

    matches = ip_pattern.findall(text)
    masked_text = ip_pattern.sub("|||IP_ADDRESS|||", text)

    return masked_text, len(matches)



In [11]:
email_tests = [
    # 普通邮箱
    "请联系 admin@example.com 获取信息。",
    # 多个邮箱
    "发送邮件到 first.user@test.com 或 second_user@domain.co.uk。",
    # 邮箱带特殊符号
    "联系 support+sales@my-domain.org 了解详情。",
    # 邮箱在句子中
    "我的邮箱是 abc.def@example.com, 请发送文件。",
    # 没有邮箱
    "这段文字中没有邮箱地址。"
]

for text in email_tests:
    masked_text, count = run_mask_emails(text)
    print(f"原文: {text}\n屏蔽后: {masked_text}\n匹配数量: {count}\n")


原文: 请联系 admin@example.com 获取信息。
屏蔽后: 请联系 |||EMAIL_ADDRESS||| 获取信息。
匹配数量: 1

原文: 发送邮件到 first.user@test.com 或 second_user@domain.co.uk。
屏蔽后: 发送邮件到 |||EMAIL_ADDRESS||| 或 |||EMAIL_ADDRESS|||。
匹配数量: 2

原文: 联系 support+sales@my-domain.org 了解详情。
屏蔽后: 联系 |||EMAIL_ADDRESS||| 了解详情。
匹配数量: 1

原文: 我的邮箱是 abc.def@example.com, 请发送文件。
屏蔽后: 我的邮箱是 |||EMAIL_ADDRESS|||, 请发送文件。
匹配数量: 1

原文: 这段文字中没有邮箱地址。
屏蔽后: 这段文字中没有邮箱地址。
匹配数量: 0



In [12]:
phone_tests = [
    # 普通美国手机号
    "请拨打 123-456-7890 联系客服。",
    # 带区号括号
    "办公室电话是 (987) 654-3210。",
    # 带国家码
    "紧急电话 +1 800 555 1212。",
    # 点号分隔
    "备用号码 555.123.4567。",
    # 没有手机号
    "这里没有电话号码。"
]

for text in phone_tests:
    masked_text, count = run_mask_phone_numbers(text)
    print(f"原文: {text}\n屏蔽后: {masked_text}\n匹配数量: {count}\n")


原文: 请拨打 123-456-7890 联系客服。
屏蔽后: 请拨打 |||PHONE_NUMBER||| 联系客服。
匹配数量: 1

原文: 办公室电话是 (987) 654-3210。
屏蔽后: 办公室电话是 (|||PHONE_NUMBER|||。
匹配数量: 1

原文: 紧急电话 +1 800 555 1212。
屏蔽后: 紧急电话 +|||PHONE_NUMBER|||。
匹配数量: 1

原文: 备用号码 555.123.4567。
屏蔽后: 备用号码 |||PHONE_NUMBER|||。
匹配数量: 1

原文: 这里没有电话号码。
屏蔽后: 这里没有电话号码。
匹配数量: 0



In [13]:
ip_tests = [
    # 普通 IP
    "服务器 IP 地址是 192.168.0.1。",
    # 边界 IP
    "网关 IP 为 255.255.255.255，子网 IP 10.0.0.0。",
    # 多个 IP
    "访问 172.16.0.1 或 8.8.8.8 获取信息。",
    # 文本中夹杂数字
    "错误码 404，用户 IP 127.0.0.1。",
    # 没有 IP
    "这段文字没有 IP 地址。"
]

for text in ip_tests:
    masked_text, count = run_mask_ips(text)
    print(f"原文: {text}\n屏蔽后: {masked_text}\n匹配数量: {count}\n")


原文: 服务器 IP 地址是 192.168.0.1。
屏蔽后: 服务器 IP 地址是 |||IP_ADDRESS|||。
匹配数量: 1

原文: 网关 IP 为 255.255.255.255，子网 IP 10.0.0.0。
屏蔽后: 网关 IP 为 |||IP_ADDRESS|||，子网 IP |||IP_ADDRESS|||。
匹配数量: 2

原文: 访问 172.16.0.1 或 8.8.8.8 获取信息。
屏蔽后: 访问 |||IP_ADDRESS||| 或 |||IP_ADDRESS||| 获取信息。
匹配数量: 2

原文: 错误码 404，用户 IP 127.0.0.1。
屏蔽后: 错误码 404，用户 IP |||IP_ADDRESS|||。
匹配数量: 1

原文: 这段文字没有 IP 地址。
屏蔽后: 这段文字没有 IP 地址。
匹配数量: 0



# 作业五

从网络上直接获取的未过滤转储文件包含大量文本，我们不希望语言模型在推理过程中重复这些内容。令人意外的是，这些训练样本本甚至可能来自看似无害的网站，比如维基百科——例如，用户在多个页面留下的评论竟颇具毒性。尽管要明确界定何为有害内容几乎不可能，但许多数据过滤流程仍会对那些主要包含有害内容的页面进行一定程度的筛选。

识别此类内容的方法有很多，包括统计黑名单中的词汇数量，或基于人工标注员给出的标签构建简单的分类器。在本次作业的这一部分，我们将重点聚焦于识别两类广泛存在的有害内容：“不适合工作场合”（NSFW）——涵盖色情、脏话或其他可能令人不适的内容——以及有毒言论（“指那些粗鲁、不尊重或无理取闹，极有可能导致他人退出讨论的语言”）。我们将利用Dolma项目提供的fasttext快速预训练模型，判断输入文本是否属于上述任一类别。这些分类器是在Jigsaw有毒评论数据集上训练而成的，该数据集包含被归类为多种不同标签的维基百科评论。

NSFW分类器可在以下网址下载：
dolma-artifacts.org/fasttext_models/jigsaw_fasttext_bigrams_20230515/jigsaw_fasttext_bigrams_nsfw_final.bin

仇恨言论分类器可在以下位置获取：
dolma-artifacts.org/fasttext_models/jigsaw_fasttext_bigrams_20230515/jigsaw_fasttext_bigrams_hatespeech_final.bin

---

使用这些模型实现一个函数，该函数接收包含页面内容的Unicode字符串，并返回一个标签（例如“有毒”、“无毒”），以及相应的置信度分数。

**问题（有害内容）：6分**

1. 编写一个函数来检测不适宜内容。
   **可交付成果**：一个函数，用于将给定字符串标记为是否包含不适宜内容，并返回包含标签和置信度分数的元组。实现适配器 `[运行NSFW分类]`，并确保其通过 `uv` 运行 `pytest -k test_classify_nsfw`。请注意，此测试仅作为基本的健全性检查，数据来源于Jigsaw语料库，但绝非用于验证您的分类器是否准确，您仍需自行进行准确性验证。

2. 编写一个函数以检测有毒言论。
   **交付物**：一个函数，用于将给定字符串标记为是否包含毒性言论，并返回一个包含标签和置信度分数的元组。实现该适配器 `[运行分类有毒言论]`，并确保其通过 `uv` 运行 `pytest -k test_classify_toxic_speech`。同样，此测试仅作为基本的正确性检查，也源自Jigsaw。

3. 你认为在将这些过滤器应用于构建训练集时，语言模型下游可能会出现哪些问题？你又该如何缓解这些问题？
   **交付物**：2至5句的回应。

4. 请对从WARC文件中提取的文本（通过您先前实现的文本提取功能）运行有害内容过滤器。随机选取20个示例进行检查，并将分类器的预测结果与您的个人判断进行对比。同时，报告任何分类器的误判情况。此外，统计有多少比例的文档被判定为有害。根据您的观察，确定在过滤过程中应采用哪些合适的分类器置信度阈值？
   **交付物**：2至5句话的回应。




下载训练好的fasttext模型

In [14]:
#NSFW分类器
!wget --no-proxy dolma-artifacts.org/fasttext_models/jigsaw_fasttext_bigrams_20230515/jigsaw_fasttext_bigrams_nsfw_final.bin

--2026-01-26 13:27:39--  http://dolma-artifacts.org/fasttext_models/jigsaw_fasttext_bigrams_20230515/jigsaw_fasttext_bigrams_nsfw_final.bin
Resolving dolma-artifacts.org (dolma-artifacts.org)... 104.21.90.110, 172.67.156.91, 2606:4700:3035::ac43:9c5b, ...
Connecting to dolma-artifacts.org (dolma-artifacts.org)|104.21.90.110|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 991753734 (946M) [application/octet-stream]
Saving to: ‘jigsaw_fasttext_bigrams_nsfw_final.bin’

jigsaw_fasttext_big 100%[===================>] 945.81M  6.32MB/s    in 2m 45s  

2026-01-26 13:30:38 (5.72 MB/s) - ‘jigsaw_fasttext_bigrams_nsfw_final.bin’ saved [991753734/991753734]



In [15]:
# 仇恨言论分类器
!wget --no-proxy dolma-artifacts.org/fasttext_models/jigsaw_fasttext_bigrams_20230515/jigsaw_fasttext_bigrams_hatespeech_final.bin

--2026-01-26 13:30:38--  http://dolma-artifacts.org/fasttext_models/jigsaw_fasttext_bigrams_20230515/jigsaw_fasttext_bigrams_hatespeech_final.bin
Resolving dolma-artifacts.org (dolma-artifacts.org)... 104.21.90.110, 172.67.156.91, 2606:4700:3037::6815:5a6e, ...
Connecting to dolma-artifacts.org (dolma-artifacts.org)|104.21.90.110|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 992245824 (946M) [application/octet-stream]
Saving to: ‘jigsaw_fasttext_bigrams_hatespeech_final.bin’

jigsaw_fasttext_big 100%[===================>] 946.28M  5.54MB/s    in 2m 42s  

2026-01-26 13:33:33 (5.84 MB/s) - ‘jigsaw_fasttext_bigrams_hatespeech_final.bin’ saved [992245824/992245824]



In [16]:
import fasttext
from typing import Tuple

# ==========================
# 1. 加载预训练模型
# ==========================

# NSFW 模型路径（本地路径）
NSFW_MODEL_PATH = "jigsaw_fasttext_bigrams_nsfw_final.bin"
# 有毒言论模型路径（本地路径）
HATE_MODEL_PATH = "jigsaw_fasttext_bigrams_hatespeech_final.bin"

# 加载 fastText 模型，注意：这一步会占用一定内存
nsfw_model = fasttext.load_model(NSFW_MODEL_PATH)
hate_model = fasttext.load_model(HATE_MODEL_PATH)

# ==========================
# 2. 定义 NSFW 分类函数
# ==========================

def classify_nsfw(text: str) -> Tuple[str, float]:
    """
    检测输入文本是否包含不适宜内容（NSFW）。

    参数:
        text (str): 待检测的 Unicode 文本字符串。

    返回:
        tuple: (标签, 置信度)
            标签: "有害" 或 "无害"
            置信度: 模型对预测标签的概率分数（0~1之间的浮点数）
    
    使用说明:
        >>> classify_nsfw("示例文本")
        ('无害', 0.95)
    """
    # fastText 的 predict 函数返回一个元组: (标签列表, 概率列表)
    labels, probs = nsfw_model.predict(text, k=1)  # 只取最可能的一个类别
    label = labels[0].replace("__label__", "")    # 去掉 fastText 特殊前缀
    confidence = probs[0]

    # 将模型标签映射为可读标签
    if label.lower() in ["nsfw", "1"]:  # 根据训练集标签可能是 "nsfw" 或 "1"
        return "有害", float(confidence)
    else:
        return "无害", float(confidence)


# ==========================
# 3. 定义有毒言论分类函数
# ==========================

def classify_toxic_speech(text: str) -> Tuple[str, float]:
    """
    检测输入文本是否包含有毒言论。

    参数:
        text (str): 待检测的 Unicode 文本字符串。

    返回:
        tuple: (标签, 置信度)
            标签: "有毒" 或 "无毒"
            置信度: 模型对预测标签的概率分数（0~1之间的浮点数）
    
    使用说明:
        >>> classify_toxic_speech("示例文本")
        ('无毒', 0.98)
    """
    # fastText 的 predict 函数返回一个元组: (标签列表, 概率列表)
    labels, probs = hate_model.predict(text, k=1)  # 只取最可能的一个类别
    label = labels[0].replace("__label__", "")     # 去掉 fastText 特殊前缀
    confidence = probs[0]

    # 将模型标签映射为可读标签
    if label.lower() in ["toxic", "1"]:  # 根据训练集标签可能是 "toxic" 或 "1"
        return "有毒", float(confidence)
    else:
        return "无毒", float(confidence)


In [17]:
text1 = "fuck"
text2 = 'hellp'
print(classify_nsfw(text1))          # ('有害', 0.87)
print(classify_toxic_speech(text2))  # ('无毒', 0.95)


('有害', 1.0000100135803223)
('无毒', 1.0000100135803223)


# 作业六


即使在按语言过滤页面并移除有害内容后，仍会有相当一部分剩余页面对语言模型训练而言质量较低。同样地，“质量”本身并不容易定义，但通过浏览Common Crawl中的示例，有助于识别低质量内容的典型情况，例如：
- 带有付费内容的页面
- 用于修复链接的占位页面
- 登录、注册或联系表单
- 主要包含非文本内容的页面，这些内容在文本提取过程中会丢失（例如：照片、视频）

Gopher论文[Rae等，2021]描述了一组简单的质量过滤器，用于从网络抓取的数据中移除低质量文本的相似简单案例。这些过滤器由易于理解的简单启发式规则组成，通常能够覆盖大量明显不适用的示例。Gopher质量过滤器包括基于文档长度、词长、符号与词的比例，以及特定英语停用词是否存在等多个标准。在本次作业中，你将实现Gopher论文[Rae等，2021]中所描述的部分过滤器。具体而言，你需要移除以下类型的文档：

- 字数少于50字或超过10万字。
- 单词平均长度超出3到10个字符的范围。
- 超过30%的行以省略号（“...”）结尾。
- 包含少于80%的至少含有一个字母字符的单词。

有关Gopher论文所使用的所有质量过滤器的完整描述，请参阅其附录A。

---

问题（gopher_quality_filters）：3分

(a) 实现（至少）上述描述的Gopher质量过滤器的一个子集。对于将文本标记化为单词，您可能会发现NLTK包很有用（特别是nltk.word_tokenize），尽管您不需要使用它。
交付物：一个函数，它接受一个字符串作为其唯一参数，并返回一个布尔值，指示文本是否通过Gopher质量过滤器。实现适配器 [run_gopher_quality_filter]。然后，确保您的过滤器通过在 uv run pytest -k test_gopher 中运行的测试。

(b) 在从WARC文件中提取的文本上运行您的基于规则的质量过滤器（通过您先前实现的文本提取函数）。查看20个随机示例，并将过滤器预测与您自己的判断进行比较。评论任何质量过滤器与您的判断不同的案例。
交付物：2-5句的回复。

In [28]:
import re
import nltk

# 如果你还没下载 punkt 分词器，需要先运行下面一行

from nltk.tokenize import word_tokenize
nltk.download('punkt')


[nltk_data] Downloading package punkt to /home/kangkang/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [31]:
nltk.download('punkt_tab')

[nltk_data] Downloading package punkt_tab to
[nltk_data]     /home/kangkang/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt_tab.zip.


True

In [ ]:

def run_gopher_quality_filter(text: str) -> bool:
    """
    Gopher 质量过滤器子集实现
    
    接受一个字符串 text，返回 True 如果文本通过过滤器，False 如果文本被过滤掉。
    
    实现的过滤规则：
    1. 字数少于50字或超过10万字 -> 不通过
    2. 单词平均长度超出 3 到 10 个字符 -> 不通过
    3. 超过30%的行以省略号结尾 -> 不通过
    4. 少于80%的单词至少含有一个字母 -> 不通过
    
    Args:
        text (str): 输入文本
    
    Returns:
        bool: True 表示通过质量过滤，False 表示不通过
    """
    
    # 规则1：字数过滤
    num_chars = len(text)
    if num_chars < 50 or num_chars > 100000:
        return False  # 字数过少或过多
    
    # 将文本按空白字符分词
    words = word_tokenize(text)
    
    # 规则2：单词平均长度过滤
    if len(words) > 0:
        avg_word_len = sum(len(w) for w in words) / len(words)
        if avg_word_len < 3 or avg_word_len > 10:
            return False  # 单词平均长度不符合要求
    
    # 规则3：省略号行比例过滤
    lines = text.splitlines()
    if lines:
        ellipsis_lines = sum(1 for line in lines if line.strip().endswith("..."))
        if ellipsis_lines / len(lines) > 0.3:
            return False  # 省略号行过多
    
    # 规则4：至少含有一个字母的单词比例过滤
    if words:
        letter_word_count = sum(1 for w in words if re.search(r"[A-Za-z]", w))
        if letter_word_count / len(words) < 0.8:
            return False  # 含字母单词比例过低
    
    # 所有规则都通过
    return True


In [32]:
# 测试 gopher_quality_filter 的简单脚本

test_cases = [
    {
        "text": "这是一个非常短的文本。",
        "expected": False,  # 字数 < 50
        "desc": "字数过少"
    },
    {
        "text": " ".join(["word"]*1000),  # 1000个单词
        "expected": True,  # 字数足够，单词长度正常
        "desc": "正常文本"
    },
    {
        "text": "word " * 10 + "..."*50 + "\n" + "word " * 10 + "...\n"*5, 
        "expected": False,  # 超过30%的行以省略号结尾
        "desc": "省略号行过多"
    },
    {
        "text": "1234 5678 91011 $$$ %%%", 
        "expected": False,  # 含字母单词比例 < 80%
        "desc": "大部分单词无字母"
    },
    {
        "text": "a " * 20 + "supercalifragilisticexpialidocious " * 5, 
        "expected": False,  # 平均单词长度 > 10
        "desc": "单词平均长度过长"
    }
]

for i, case in enumerate(test_cases):
    result = gopher_quality_filter(case["text"])
    print(f"测试案例 {i+1} ({case['desc']}): {'通过' if result else '不通过'}, 期望: {'通过' if case['expected'] else '不通过'}")


测试案例 1 (字数过少): 不通过, 期望: 不通过
测试案例 2 (正常文本): 通过, 期望: 通过
测试案例 3 (省略号行过多): 不通过, 期望: 不通过
测试案例 4 (大部分单词无字母): 不通过, 期望: 不通过
测试案例 5 (单词平均长度过长): 通过, 期望: 不通过


# 作业七 

质量分类器

现在让我们超越Gopher规则所捕获的简单句法标准，考虑质量评估。语言模型训练并不是根据内容质量进行排名的首选方法。特别是，文本质量也是信息检索中的一个基本挑战。搜索引擎利用网页链接结构的经典信号之一是：高质量页面倾向于链接到其他高质量页面 [Page et al., 1998]。OpenAI在构建WebText时使用了类似的见解，GPT-2 [Radford et al., 2019] 训练的数据集包括那些在Reddit评论中获得高于最低“karma”阈值的页面。另一个选择是使用Wikipedia作为高质量链接的来源，因为从Wikipedia页面链接的外部来源往往是可信的 [Touvron et al., 2023]。

虽然使用受控来源通常会导致高质量内容，但结果数据集最终对于今天的标准来说太小了（OpenWebText有40GB的文本，而Pile比它大20倍）。一种方法是使用这些参考页面作为正面例子，从Common Crawl中抽取的（随机）页面作为负面例子，并训练一个fastText分类器。这个分类器会给出质量分数，您可以使用它来过滤掉整个Common Crawl中的页面。设置质量阈值会在精确度和召回率之间进行权衡。

在本次作业中，您将构建一个质量分类器。为了方便，我们已经从最近的Wikipedia转储中提取了参考页面的URL，并将其放置在Together集群的 /data/wiki/enwiki-20240420-extracted_urls.txt.gz。该文件包含截至2024年4月在英语维基百科页面上找到的43,500个外部链接的列表，但我们希望您订阅这些URL以获得用于训练您分类器的“高质量”文本的正面例子。请注意，这些正面例子可能仍包含不理想的内容，因此应用您构建的其他原语可能是有用的。

（例如，语言识别、过滤规则等）以进一步提高其质量。给定一个URL文件，您可以使用wget以WARC格式抓取其内容：

```
wget --timeout=5 \
  -i subsampled_positive_urls.txt \
  --warc-file=subsampled_positive_urls.warc \
  -O /dev/null
```

问题（quality_classifier）：15分

(a) 训练一个质量分类器，给定文本，返回一个数值质量分数。
交付物：一个用于下一个子问题的可用质量分类器。

(b) 编写一个函数，将页面标记为高质量或低质量，并在标签中提供置信度分数。
交付物：一个函数，它接受一个字符串作为其唯一参数，并返回一个带有标签（高质量或非高质量）和置信度分数的对。实现适配器 [run_classify_quality]。作为基本检查，确保它通过运行 uv run pytest -k test_classify_quality 正确分类我们提供的两个示例。

In [36]:
!wget https://downloads.cs.stanford.edu/nlp/data/nfliu/cs336-spring-2024/assignment4/enwiki-20240420-extracted_urls.txt.gz


--2026-01-26 16:33:15--  https://downloads.cs.stanford.edu/nlp/data/nfliu/cs336-spring-2024/assignment4/enwiki-20240420-extracted_urls.txt.gz
Connecting to 127.0.0.1:7890... connected.
Proxy request sent, awaiting response... 200 OK
Length: 1055989383 (1007M) [application/octet-stream]
Saving to: ‘enwiki-20240420-extracted_urls.txt.gz.1’

enwiki-20240420-ext 100%[===================>]   1007M  3.19MB/s    in 5m 48s  

2026-01-26 16:39:23 (2.90 MB/s) - ‘enwiki-20240420-extracted_urls.txt.gz.1’ saved [1055989383/1055989383]



In [37]:
!gunzip enwiki-20240420-extracted_urls.txt.gz

使用以下官方给的命令，并发太大，消耗内存导致wsl崩溃！几乎所有的崩溃都和这个原因有关
所以我使用以下代码下载
```bash
wget --timeout=5 \
  -i enwiki-20240420-extracted_urls.txt \
  --warc-file=subsampled_positive_urls.warc \
  -O /dev/null
```

In [1]:
import requests
from requests.adapters import HTTPAdapter
from urllib3.util.retry import Retry
from bs4 import BeautifulSoup
import json
from concurrent.futures import ThreadPoolExecutor, as_completed
from tqdm import tqdm  # 用于显示进度条
import time
import os

# =====================
# 配置参数
# =====================
MAX_WORKERS = 12       # 并发线程数，WSL 推荐 4–6，太高容易崩
TIMEOUT = 20           # 网络请求超时时间（秒）
MIN_TEXT_LEN = 100     # 文本最小长度，小于该长度的网页会被丢弃
OUTPUT = "wiki_train.jsonl"  # 训练数据输出文件，每行 JSON
DONE_FILE = "done_urls.txt"  # 已处理 URL 文件，用于断点续跑
URL_FILE = "enwiki-20240420-extracted_urls.txt"  # 待抓取 URL 列表

HEADERS = {
    "User-Agent": "WikiDataCollector/1.0 (training)"  # 避免被网站封禁
}

# =====================
# 读取已完成 URL（断点续跑）
# =====================
done_urls = set()
if os.path.exists(DONE_FILE):
    # 如果 done 文件存在，就把所有 URL 加入 set
    # 避免重复抓取
    with open(DONE_FILE, "r", encoding="utf-8") as f:
        for line in f:
            done_urls.add(line.strip())

# =====================
# URL 流生成器（跳过已完成 URL）
# =====================
def url_stream(path):
    """
    按行读取 URL 文件，并跳过已经完成的 URL
    使用生成器可以避免一次性加载大文件，节约内存
    """
    with open(path, "r", encoding="utf-8") as f:
        for line in f:
            url = line.strip()
            if url and url not in done_urls:
                yield url

# =====================
# 配置 requests Session（复用连接 + 重试）
# =====================
session = requests.Session()
# 设置重试策略，遇到指定 HTTP 错误码自动重试
retries = Retry(total=2, backoff_factor=1, status_forcelist=[429,500,502,503,504])
# 配置连接池大小，避免线程阻塞
adapter = HTTPAdapter(pool_connections=MAX_WORKERS*2, pool_maxsize=MAX_WORKERS*2, max_retries=retries)
session.mount("http://", adapter)
session.mount("https://", adapter)

# =====================
# 网页抓取与文本清洗函数
# =====================
def fetch_and_process(url):
    """
    对单个 URL 抓取网页、提取纯文本
    返回一个字典 {"url", "length", "text"}，否则返回 None
    """
    try:
        # 发起 GET 请求
        r = session.get(url, headers=HEADERS, timeout=TIMEOUT)

        # 非 200 响应直接丢弃
        if r.status_code != 200:
            return None

        # 非 HTML 内容丢弃
        if "text/html" not in r.headers.get("Content-Type", ""):
            return None

        # 使用 BeautifulSoup 清洗 HTML
        soup = BeautifulSoup(r.text, "html.parser")  # ⚠ 用 html.parser 更稳，不依赖 lxml

        # 移除脚本、样式、noscript 标签
        for tag in soup(["script", "style", "noscript"]):
            tag.decompose()

        # 提取纯文本，并压缩空格
        text = " ".join(soup.get_text(separator=" ").split())

        # 文本过短丢弃
        if len(text) < MIN_TEXT_LEN:
            return None

        # 返回字典，用于写入 JSONL
        return {"url": url, "length": len(text), "text": text}

    except Exception:
        # 抓取或解析异常直接返回 None，不中断整个流程
        return None

# =====================
# 主流程
# =====================
def main():
    # 生成待抓取 URL 列表
    urls = list(url_stream(URL_FILE))
    total = len(urls)  # 总 URL 数量，用于进度条

    # 打开输出文件和断点续跑文件
    with open(OUTPUT, "a", encoding="utf-8") as out, \
         open(DONE_FILE, "a", encoding="utf-8") as done_out, \
         ThreadPoolExecutor(max_workers=MAX_WORKERS) as executor:

        futures = []  # 存储线程任务
        pbar = tqdm(total=total, desc="Processing URLs")  # 进度条

        for url in urls:
            # 提交线程池任务
            futures.append(executor.submit(fetch_and_process, url))

            # 控制 futures 批量处理，防止无限增长占用内存
            if len(futures) >= MAX_WORKERS*4:
                for future in as_completed(futures):
                    result = future.result()
                    if result:
                        # 写入 JSONL 输出文件
                        out.write(json.dumps(result) + "\n")
                        out.flush()  # 确保写入磁盘

                        # 记录已完成 URL
                        done_out.write(result["url"] + "\n")
                        done_out.flush()

                    # 更新进度条
                    pbar.update(1)
                futures.clear()  # 清空批量 futures

        # 处理剩余未完成的 futures
        for future in as_completed(futures):
            result = future.result()
            if result:
                out.write(json.dumps(result) + "\n")
                out.flush()
                done_out.write(result["url"] + "\n")
                done_out.flush()
            pbar.update(1)

        pbar.close()  # 关闭进度条

# =====================
# 脚本入口
# =====================
if __name__ == "__main__":
    main()


Processing URLs:   0%|          | 191/43579680 [08:03<31399:58:11,  2.59s/it] 

KeyboardInterrupt: 

但是下载还是太慢了，最后直接下载数据集

In [2]:
!wget https://dumps.wikimedia.org/enwiki/latest/enwiki-latest-pages-articles.xml.bz2

--2026-01-26 21:06:40--  https://dumps.wikimedia.org/enwiki/latest/enwiki-latest-pages-articles.xml.bz2
Connecting to 127.0.0.1:7890... connected.
Proxy request sent, awaiting response... 200 OK
Length: 24796216934 (23G) [application/octet-stream]
Saving to: ‘enwiki-latest-pages-articles.xml.bz2’

test-pages-articles   6%[>                   ]   1.43G  4.03MB/s    eta 97m 41s^C


上面是24G版本，所占内存太大，如果只是体验流程可以下载以下小版本

In [46]:
!wget  https://dumps.wikimedia.org/enwiki/latest/enwiki-latest-pages-articles1.xml-p1p41242.bz2

--2026-01-27 12:34:24--  https://dumps.wikimedia.org/enwiki/latest/enwiki-latest-pages-articles1.xml-p1p41242.bz2
Connecting to 127.0.0.1:7890... connected.
Proxy request sent, awaiting response... 200 OK
Length: 293742434 (280M) [application/octet-stream]
Saving to: ‘enwiki-latest-pages-articles1.xml-p1p41242.bz2’

enwiki-latest-pages 100%[===================>] 280.13M  3.62MB/s    in 78s     

2026-01-27 12:35:54 (3.61 MB/s) - ‘enwiki-latest-pages-articles1.xml-p1p41242.bz2’ saved [293742434/293742434]



In [47]:
!wget https://dumps.wikimedia.org/enwiki/latest/enwiki-latest-pages-articles10.xml-p4045403p5399366.bz2

--2026-01-27 12:36:52--  https://dumps.wikimedia.org/enwiki/latest/enwiki-latest-pages-articles10.xml-p4045403p5399366.bz2
Connecting to 127.0.0.1:7890... connected.
Proxy request sent, awaiting response... 200 OK
Length: 590499085 (563M) [application/octet-stream]
Saving to: ‘enwiki-latest-pages-articles10.xml-p4045403p5399366.bz2’

enwiki-latest-pages 100%[===================>] 563.14M  4.31MB/s    in 2m 17s  

2026-01-27 12:39:27 (4.11 MB/s) - ‘enwiki-latest-pages-articles10.xml-p4045403p5399366.bz2’ saved [590499085/590499085]



In [48]:
!wget https://dumps.wikimedia.org/enwiki/latest/enwiki-latest-pages-articles11.xml-p5399367p6899366.bz2

--2026-01-27 12:39:29--  https://dumps.wikimedia.org/enwiki/latest/enwiki-latest-pages-articles11.xml-p5399367p6899366.bz2
Connecting to 127.0.0.1:7890... connected.
Proxy request sent, awaiting response... 200 OK
Length: 565952787 (540M) [application/octet-stream]
Saving to: ‘enwiki-latest-pages-articles11.xml-p5399367p6899366.bz2’

enwiki-latest-pages 100%[===================>] 539.73M  4.11MB/s    in 2m 16s  

2026-01-27 12:42:01 (3.95 MB/s) - ‘enwiki-latest-pages-articles11.xml-p5399367p6899366.bz2’ saved [565952787/565952787]



下载好之后要使用wikiextractor 进行 格式转换，这个项目支持的python版本为3.9，本项目不支持，要另开一个环境。

打开vscode新窗口（或者别的软件也行），在别的目录下执行：

克隆项目
```bash
git clone https://github.com/attardi/wikiextractor.git
···

```bash
cd  wikiextractor
```

创建环境
```bash
uv venv --python=3.9

. .venv/bin/activate
```




使用该项目 记得将输入的地址和输出的地址进行替换，这里是作者的目录

```bash

python -m wikiextractor.WikiExtractor \
  /home/kangkang/my_project/CS336-Chinese-co-construction/coursework/Assignment4_Data/cs336_data/enwiki-latest-pages-articles1.xml-p1p41242.bz2\ 
  --output /home/kangkang/my_project/CS336-Chinese-co-construction/coursework/Assignment4_Data/cs336_data/extracted/output1 \
  --json \
  --bytes 10M \
  --processes 8

python -m wikiextractor.WikiExtractor \
  /home/kangkang/my_project/CS336-Chinese-co-construction/coursework/Assignment4_Data/cs336_data/enwiki-latest-pages-articles10.xml-p4045403p5399366.bz2\
  --output /home/kangkang/my_project/CS336-Chinese-co-construction/coursework/Assignment4_Data/cs336_data/extracted/output12 \
  --json \
  --bytes 10M \
  --processes 8


python -m wikiextractor.WikiExtractor \
  /home/kangkang/my_project/CS336-Chinese-co-construction/coursework/Assignment4_Data/cs336_data/enwiki-latest-pages-articles11.xml-p5399367p6899366.bz2\
  --output /home/kangkang/my_project/CS336-Chinese-co-construction/coursework/Assignment4_Data/cs336_data/extracted/output13 \
  --json \
  --bytes 10M \
  --processes 8

```


In [1]:
from filter import run_mask_emails, run_mask_phone_numbers, run_mask_ips, run_gopher_quality_filter

使用之前写的函数进行清洗

In [35]:
import json
import sys
from pathlib import Path
from typing import Generator, Optional,Callable
from tqdm import tqdm

def read_extracted_files(extracted_dir: str) -> Generator[dict, None, None]:
    """
    读取所有提取的JSONL文件
    """
    extracted_path = Path(extracted_dir)
    
    # 遍历所有子目录
    for wiki_file in extracted_path.rglob('wiki_*'):
         
         try:
             with open(wiki_file, 'r', encoding='utf-8') as f:
                 for line in f:
                     line = line.strip()
                     if not line:
                         continue
                     try:
                         article = json.loads(line)
                         yield article #惰性迭代，每次只返回一个article
                     except json.JSONDecodeError:
                         continue
         except Exception as e:
            print(f"Error reading {wiki_file}: {e}")
            continue

def desensitize_text(text: str) -> str:
    """
    对文本进行脱敏处理
    """
    # 使用已有的脱敏函数
    text, _ = run_mask_emails(text)
    text, _ = run_mask_phone_numbers(text)
    text, _ = run_mask_ips(text)
    
    return text


def check_gopher_quality_rules(text: str) -> bool:
    """
    检查文本是否符合Gopher质量规则
    - 词数大于50 小于100000
    - 平均词长3-10
    - 省略号占比小于30%
    """
    return run_gopher_quality_filter(text)

def process_text(text: str) -> Optional[str]:
    """
    处理单个文本：脱敏 + 质量过滤
    
    Returns:
        处理后的文本，如果不符合质量要求则返回None
    """
    if not text or not text.strip():
        return None
    
    # 1. 脱敏
    text = desensitize_text(text)
    
    # 2. Gopher质量规则过滤
    if not check_gopher_quality_rules(text):
        return None
    
    # 3. 清理：压缩空格，转换为单行（fastText要求）
    text = ' '.join(text.split())
    
    return text


def save_fasttext_format(texts: list[str], label: str, output_path: str,
                         callback: Optional[Callable[[], None]] = None):
    with open(output_path, 'w', encoding='utf-8') as f:
        for text in texts:
            f.write(f"__label__{label} {text}\n")
            if callback:          # 每条写完后触发
                callback()

from tqdm import tqdm

def main() -> None:
    extracted_dir = "extracted"
    output_path   = "wiki_positive_samples.txt"

    processed_texts = []
    stats = {"total": 0, "empty": 0, "quality": 0}

    # 用 tqdm 包装迭代器，同时把统计写进 postfix
    with tqdm(desc="Processing", unit="art") as pbar:
        for article in read_extracted_files(extracted_dir):
            stats["total"] += 1
            text = article.get("text", "")

            if not text.strip():
                stats["empty"] += 1
                pbar.set_postfix(skipped_empty=stats["empty"], skipped_quality=stats["quality"])
                pbar.update()
                continue

            processed = process_text(text)
            if processed is None:
                stats["quality"] += 1
                pbar.set_postfix(skipped_empty=stats["empty"], skipped_quality=stats["quality"])
                pbar.update()
                continue

            processed_texts.append(processed)
            pbar.update()

            # 每 1 k 条刷新一次计数
            if len(processed_texts) % 1000 == 0:
                pbar.set_postfix(valid=len(processed_texts))

    # 保存阶段也包一层进度条
    if processed_texts:
        with tqdm(total=len(processed_texts), desc="Saving FastText") as pbar:
            save_fasttext_format(processed_texts, label="hq", output_path=output_path,
                                 callback=lambda: pbar.update())  # 若 save 函数支持回调
    else:
        save_fasttext_format(processed_texts, label="hq", output_path=output_path)

    print(f"Done! 总文章: {stats['total']}, 有效: {len(processed_texts)}, "
          f"空文本: {stats['empty']}, 质量过滤: {stats['quality']}")


if __name__ == "__main__":
    main()


Processing: 780806art [22:42, 573.09art/s, skipped_empty=420664, skipped_quality=112507] 
Saving FastText: 100%|██████████| 247635/247635 [00:10<00:00, 24427.32it/s]


Done! 总文章: 780806, 有效: 247635, 空文本: 420664, 质量过滤: 112507


In [ ]:
%%writefile cc2train.py

#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
从 Common Crawl WARC 中提取负样本（lq）并保存为 FastText 格式
新增：文件级 + 记录级 双进度条，实时跳过统计，支持 Ctrl-C 中断
"""
import gzip
import signal
import sys
import random
from pathlib import Path
from typing import Optional, List
from tqdm import tqdm
from warcio.archiveiterator import ArchiveIterator

# 如果脚本在子目录，把项目根加入 PYTHONPATH
sys.path.insert(0, str(Path(__file__).parent.parent))
from filter import (
    run_mask_emails,
    run_mask_ips,
    run_mask_phone_numbers,
    run_gopher_quality_filter,
    run_extract_text_from_html_bytes,
)

# -------------------- 工具函数 -------------------- #
def desensitize_text(text: str) -> str:
    """邮箱 / 电话 / IP 脱敏"""
    text, _ = run_mask_emails(text)
    text, _ = run_mask_phone_numbers(text)
    text, _ = run_mask_ips(text)
    return text


def extract_text_from_warc_record(record) -> Optional[str]:
    """
    从单个 WARC 记录中提取纯文本
    只处理 HTTP 200 且 Content-Type 为 text/html 的响应
    """
    if record.rec_type != 'response':
        return None
    content_type = record.http_headers.get_header('Content-Type', '')
    if 'text/html' not in content_type.lower():
        return None

    try:
        html_bytes = record.content_stream().read()
        text = run_extract_text_from_html_bytes(html_bytes)
        if text and len(text.strip()) > 100:          # 太短不要
            return text
    except Exception:
        return None
    return None


def process_text(text: str) -> Optional[str]:
    """脱敏 + 单行化"""
    if not text or not text.strip():
        return None
    text = desensitize_text(text)
    return ' '.join(text.split())                     # 去多余空白/换行


# -------------------- 核心提取 -------------------- #
def extract_negative_samples_from_warc(
    warc_paths: List[str],
    target_count: int,
    random_seed: int = 42,
) -> List[str]:
    """
    多文件 WARC → 提取负样本
    文件级 + 记录级 双进度条，实时显示跳过原因
    """
    random.seed(random_seed)
    results: List[str] = []

    # 文件级进度条
    pbar_files = tqdm(warc_paths, desc="WARC文件", unit="file", position=0)
    skipped  = {"empty": 0, "quality": 0}

    for warc_path in pbar_files:
        if len(results) >= target_count:
            pbar_files.write("已达到目标数量，提前结束")
            break
        if not Path(warc_path).exists():
            skipped["file_not_found"] = skipped.get("file_not_found", 0) + 1
            continue

        # 先扫一遍总记录数，用于记录级进度条
        with gzip.open(warc_path, 'rb') as f:
            total_records = sum(1 for _ in ArchiveIterator(f))
        pbar_files.set_postfix(file=Path(warc_path).name, total_records=total_records)

        # 第二遍真正提取
        with gzip.open(warc_path, 'rb') as f:
            for record in tqdm(
                ArchiveIterator(f),
                total=total_records,
                desc="记录",
                unit="rec",
                position=1,
                leave=False,
            ):
                if len(results) >= target_count:
                    break
                text = extract_text_from_warc_record(record)
                if text is None:
                    skipped["empty"] += 1
                    continue
                processed = process_text(text)
                if processed is None:
                    skipped["quality"] += 1
                    continue
                results.append(processed)

        pbar_files.set_postfix(
            valid=len(results), skipped_empty=skipped["empty"], skipped_quality=skipped["quality"]
        )

    return results


# -------------------- 辅助函数 -------------------- #
def count_positive_samples(positive_file: str) -> int:
    """统计正样本行数（__label__hq 开头）"""
    try:
        with open(positive_file, 'r', encoding='utf-8') as f:
            return sum(1 for line in f if line.strip().startswith('__label__hq'))
    except FileNotFoundError:
        return 0


def find_warc_files(datasets_dir: Path) -> List[str]:
    """扫描 datasets 目录下 *.warc.gz，排除 .wet.gz"""
    return sorted(str(f) for f in datasets_dir.glob("*.warc.gz") if not f.name.endswith(".wet.gz"))


# -------------------- 主入口 -------------------- #
def main():
    script_dir    = Path(__file__).parent
    datasets_dir  = Path(__file__).parent
    positive_file = str(script_dir / "wiki_positive_samples.txt")
    output_path   = str(script_dir / "wiki_negative_samples.txt")

    # 1. 确定目标数量
    target_count  = count_positive_samples(positive_file) or 10_000
    print("正样本数量为",target_count)
    warc_paths    = find_warc_files(datasets_dir)
    if not warc_paths:
        print("未找到任何 WARC 文件，程序结束")
        return

    # 2. 提取负样本
    print(f"目标负样本数量：{target_count}")
    negative_texts = extract_negative_samples_from_warc(warc_paths, target_count)

    # 3. 保存
    if negative_texts:
        with open(output_path, 'w', encoding='utf-8') as f:
            for text in negative_texts:
                f.write(f"__label__lq {text}\n")
        print(f"已保存 {len(negative_texts)} 条负样本 → {output_path}")
    else:
        print("没有提取到任何负样本")


if __name__ == "__main__":
    # 支持 Ctrl-C 中断
    signal.signal(signal.SIGINT, lambda *_: (print("\n用户中断，退出"), sys.exit(0)))
    main()

Overwriting cc2train.py


之前下载的数据集只是一部分，数量不够，这里多下几个

In [18]:
!wget https://data.commoncrawl.org/crawl-data/CC-MAIN-2025-51/segments/1764871306713.64/warc/CC-MAIN-20251204191828-20251204221828-00000.warc.gz

--2026-01-27 21:47:04--  https://data.commoncrawl.org/crawl-data/CC-MAIN-2025-51/segments/1764871306713.64/warc/CC-MAIN-20251204191828-20251204221828-00000.warc.gz
Connecting to 127.0.0.1:7890... connected.
Proxy request sent, awaiting response... 200 OK
Length: 914552144 (872M) [application/octet-stream]
Saving to: ‘CC-MAIN-20251204191828-20251204221828-00000.warc.gz’

CC-MAIN-20251204191 100%[===================>] 872.18M  2.20MB/s    in 5m 17s  

2026-01-27 21:52:22 (2.75 MB/s) - ‘CC-MAIN-20251204191828-20251204221828-00000.warc.gz’ saved [914552144/914552144]



In [19]:
!wget https://data.commoncrawl.org/crawl-data/CC-MAIN-2025-51/segments/1764871306713.64/warc/CC-MAIN-20251204191828-20251204221828-00001.warc.gz

--2026-01-27 21:52:56--  https://data.commoncrawl.org/crawl-data/CC-MAIN-2025-51/segments/1764871306713.64/warc/CC-MAIN-20251204191828-20251204221828-00001.warc.gz
Connecting to 127.0.0.1:7890... connected.
Proxy request sent, awaiting response... 200 OK
Length: 878685677 (838M) [application/octet-stream]
Saving to: ‘CC-MAIN-20251204191828-20251204221828-00001.warc.gz’

CC-MAIN-20251204191 100%[===================>] 837.98M  1.52MB/s    in 7m 50s  

2026-01-27 22:00:47 (1.78 MB/s) - ‘CC-MAIN-20251204191828-20251204221828-00001.warc.gz’ saved [878685677/878685677]



In [20]:
!wget https://data.commoncrawl.org/crawl-data/CC-MAIN-2025-51/segments/1764871306713.64/warc/CC-MAIN-20251204191828-20251204221828-00002.warc.gz

--2026-01-27 22:00:54--  https://data.commoncrawl.org/crawl-data/CC-MAIN-2025-51/segments/1764871306713.64/warc/CC-MAIN-20251204191828-20251204221828-00002.warc.gz
Connecting to 127.0.0.1:7890... connected.
Proxy request sent, awaiting response... 200 OK
Length: 907176983 (865M) [application/octet-stream]
Saving to: ‘CC-MAIN-20251204191828-20251204221828-00002.warc.gz’

CC-MAIN-20251204191 100%[===================>] 865.15M  1.51MB/s    in 9m 49s  

2026-01-27 22:10:44 (1.47 MB/s) - ‘CC-MAIN-20251204191828-20251204221828-00002.warc.gz’ saved [907176983/907176983]



In [21]:

!python cc2train.py


正样本数量为 247635
目标负样本数量：247635
WARC文件:   0%| | 0/4 [00:19<?, ?file/s, file=CC-MAIN-20250417135010-20250417165
记录: 100%|███████████████████████████▉| 83361/83473 [01:20<00:00, 1000.18rec/s]
WARC文件:  25%|▎| 1/4 [01:54<04:57, 99.14s/file, file=CC-MAIN-20251204191828-202  
记录: 100%|███████████████████████████▉| 66381/66478 [01:07<00:00, 1379.53rec/s]
WARC文件:  50%|▌| 2/4 [03:17<03:00, 90.03s/file, file=CC-MAIN-20251204191828-202  
记录: 100%|███████████████████████████▉| 65112/65224 [01:15<00:00, 1045.52rec/s]
WARC文件:  75%|▊| 3/4 [04:47<01:30, 90.18s/file, file=CC-MAIN-20251204191828-202  
记录: 100%|████████████████████████████▉| 65817/65848 [01:06<00:00, 987.54rec/s]
WARC文件: 100%|█| 4/4 [05:53<00:00, 88.48s/file, skipped_empty=193185, skipped_q  
已保存 87838 条负样本 → /home/kangkang/my_project/CS336-Chinese-co-construction/coursework/Assignment4_Data/cs336_data/wiki_negative_samples.txt


接下来是模型的训练

In [26]:
%%writefile train_fasttext.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
训练 FastText 质量分类器（二分类：hq vs lq）
步骤：合并 → 打乱 → 划分 → 训练 → 保存模型
"""
import fasttext
import random
from pathlib import Path
from typing import Optional, Tuple, List
from tqdm import tqdm

# ---------- 1. 合并 + 打乱 ---------- #
def merge_and_shuffle_samples(
    positive_file: str,
    negative_file: str,
    output_file: str,
    shuffle: bool = True,
    random_seed: int = 42
) -> int:
    """
    合并正负样本并打乱顺序，返回总样本数
    注意：必须混排，否则 batch 内类别单一，训练失效
    """
    random.seed(random_seed)
    all_samples: List[str] = []

    # 读正样本
    for line in tqdm(open(positive_file, 'r', encoding='utf-8'), desc="读取正样本"):
        line = line.strip()
        if line.startswith('__label__hq'):
            all_samples.append(line)

    # 读负样本
    for line in tqdm(open(negative_file, 'r', encoding='utf-8'), desc="读取负样本"):
        line = line.strip()
        if line.startswith('__label__lq'):
            all_samples.append(line)

    if shuffle:
        random.shuffle(all_samples)

    # 写合并文件
    with open(output_file, 'w', encoding='utf-8') as f:
        for line in tqdm(all_samples, desc="写入合并文件"):
            f.write(line + '\n')

    return len(all_samples)

# ---------- 2. 训练 / 验证划分 ---------- #
def split_train_val(
    merged_file: str,
    train_file: str,
    val_file: str,
    val_ratio: float = 0.1,
    random_seed: int = 42
) -> Tuple[int, int]:
    """按 val_ratio 划分训练集与验证集，返回 (训练条数, 验证条数)"""
    random.seed(random_seed)
    lines = [ln.strip() for ln in open(merged_file, 'r', encoding='utf-8') if ln.strip()]
    random.shuffle(lines)
    val_size = int(len(lines) * val_ratio)
    with open(val_file, 'w', encoding='utf-8') as f:
        for ln in tqdm(lines[:val_size], desc="写验证集"):
            f.write(ln + '\n')
    with open(train_file, 'w', encoding='utf-8') as f:
        for ln in tqdm(lines[val_size:], desc="写训练集"):
            f.write(ln + '\n')
    return len(lines) - val_size, val_size

# ---------- 3. 训练 + 评估 ---------- #
def train_quality_classifier(
    train_path: str,
    model_path: str,
    val_path: Optional[str] = None,
    **kwargs
) -> fasttext.FastText:
    """
    训练 fastText 分类器并保存
    默认参数已针对中文/英文维基质量过滤调优，可外部覆盖
    """
    defaults = dict(
        lr=0.1,
        epoch=25,
        wordNgrams=2,
        dim=100,
        loss='softmax',
        minCount=5,
        bucket=2_000_000,
        thread=8,          # 多线程加速
    )
    defaults.update(kwargs)
    model = fasttext.train_supervised(input=train_path, **defaults)
    model.save_model(model_path)
    if val_path and Path(val_path).exists():
        n, p, r = model.test(val_path)
        print(f"验证集  样本数:{n}  precision:{p:.4f}  recall:{r:.4f}  F1:{2*p*r/(p+r):.4f}")
    return model

# ---------- 4. 主流程 ---------- #
def main() -> None:
    script_dir = Path(__file__).parent
    pos_file   = script_dir / "wiki_positive_samples.txt"
    neg_file   = script_dir / "wiki_negative_samples.txt"
    merged     = script_dir / "quality_merged.txt"
    train_f    = script_dir / "quality_train.txt"
    val_f      = script_dir / "quality_val.txt"
    model_bin  = script_dir / "quality_classifier.bin"

    # 1. 合并 & 打乱
    total = merge_and_shuffle_samples(str(pos_file), str(neg_file), str(merged))
    if total == 0:
        print("没有有效样本，程序结束")
        return

    # 2. 划分
    train_n, val_n = split_train_val(str(merged), str(train_f), str(val_f))

    # 3. 训练
    print(f"开始训练：训练集 {train_n} 条，验证集 {val_n} 条")
    train_quality_classifier(str(train_f), str(model_bin), str(val_f))

    print(f"模型已保存至 {model_bin}")


if __name__ == "__main__":
    main()

Writing train_fasttext.py


In [27]:
!python train_fasttext.py

读取正样本: 247635it [00:03, 67165.52it/s]
读取负样本: 87838it [00:03, 26537.07it/s]
写训练集: 100%|██████████████████████| 301926/301926 [00:04<00:00, 61734.01it/s]
开始训练：训练集 301926 条，验证集 33547 条
Read 266M words
Number of words:  1737920
Number of labels: 2
Progress: 100.0% words/sec/thread: 1584252 lr:  0.000000 avg.loss:  0.003925 ETA:   0h 0m 0s% words/sec/thread: 1445540 lr:  0.095204 avg.loss:  0.018011 ETA:   0h 9m 7s ETA:   0h 9m 4s 0.081651 avg.loss:  0.008626 ETA:   0h 7m25s 0.007569 ETA:   0h 6m49s 0.005908 ETA:   0h 5m23s55s 0.005064 ETA:   0h 3m55sm40s 0.018191 avg.loss:  0.004244 ETA:   0h 1m35s ETA:   0h 1m33sh 0m12s
验证集  样本数:33547  precision:0.9990  recall:0.9990  F1:0.9990
模型已保存至 /home/kangkang/my_project/CS336-Chinese-co-construction/coursework/Assignment4_Data/cs336_data/quality_classifier.bin


In [33]:
import fasttext
from typing import Any

def run_classify_quality(text: str) -> tuple[Any, float]:
    model = fasttext.load_model('quality_classifier.bin')
    perdiction = model.predict(text.replace('\n', ' '))

    label = perdiction[0][0]   
    score = perdiction[1][0]

    #去掉 __label__前缀
    quality_detect = label.replace('__label__', '')
    return (quality_detect, score)

In [35]:
with open('wiki_positive_samples.txt','r') as f:
    for line in f:
        print(line)
        print(run_classify_quality(line.replace('__label__hq ','')))
        break
with open('wiki_negative_samples.txt','r') as f:
    for line in f:
        print(line)
        print(run_classify_quality(line.replace('__label__hq ','')))
        break

__label__hq The West Vancouver Police Department (WVPD; ) is the municipal police force for the district of West Vancouver, British Columbia, Canada. Operations. The West Vancouver Police respond to calls for service using one of four Patrol Teams. Each Team is staffed by several Constables, a Sergeant (Road Supervisor), and with oversight from a Staff Sergeant (Watch Commander). The patrol division is augmented by several support sections including, Traffic, Community Policing / School Liaison, Records, Surveillance, Intel., and Victim Services. Serious crimes are investigated by Detectives on the Major Investigations Team. Homicides that occur in the District of West Vancouver are investigated by the Integrated Homicide Investigation Team which is supplemented by a seconded WVPD member. To further support the concept of seamless policing within the Lower Mainland region, several WVPD members are seconded to various other integrated municipal/provincial RCMP units. These include; the 

1.0000100135803223 并不是“算错了”，而是 fastText 内部数值精度与近似计算 导致的 微溢出（slightly > 1.0），在二分类里很常见，不影响使用